### ISSUE TITLE
Best Practices for logging image predictions while referencing only the original version-controlled artifact media files[Q]

### ISSUE THREAD

**Issue by JacobARose** [wandb/client#2216]
This references tssweeney's PR here: #1438

### Overview

I have a question about the best way to implement this as a workflow. I'd like to repeatedly log predictions on a version-controlled test set of images with class labels, while:

A. Ensuring that the prediction tables correctly associate any subsequent predictions on sample X with unique ID X_id across multiple evaluations of different models.

B. Ensure that the image file corresponding with a particular prediction is correctly uploaded essentially as a reference to the original server-side version-controlled image file, instead of uploading a duplicate image file upon every round of predictions.

### Example artifacts layout:
```

artifact 1  
     /media  
            /image1  
            /image2  
            ...

artifact 2
    /tables
            /test_prediction_table.json
    /media
            /image1
            /image2
            ...
```

In the artifacts above, artifact 2 contains a table with one column referencing a hosted image file, and some other columns each representing a prediction on that image. As it stands now, every time I upload a prediction table, it results in a situation like this, where each separate artifact contains its own copy of the same images. What I would like is for the test_prediction_table.json to only reference the images in artifact 1/media, and keep the media subdir for artifact 2 empty.

tldr; I can successfully track predictions per-image across multiple evaluations logged at different times, but do not yet have a clear method for preventing the redundant duplication of raw image data upon every new log event.

---------------------

### Current possibilities

I guess there are a few ways I could imagine this working.

1. My default strategy, performing predictions normally then wrapping the image in a wandb.Image constructor every time, allows minimal refactoring of my workflow. However, this seems unlikely to be well-suited for perfectly reproducing images automatically in a way that links them to any pre-existing server-side artifacts. At the moment, it appears this is uploading an entire copy of every image for each round of predictions. While I am successfully able to compare predictions across models based on always including a common unique image ID, the value of this is greatly diminished if Im rapidly using up all storage space.
2. Using artifact.get for each image individually seems like it would come with some serious computational and software design-related inefficiencies, in addition to requiring users to refactor their data loading pipelines around WandB's artifact API in a more fundamental way than simply using it to download a collection of files to a local directory.
3. A user must also consider when and where to apply standard image preprocessing to an downloaded image being used for prediction, as this isnt always (and shouldnt be required to be) already applied to version-controlled images.
4. 2 of the strategies mentioned in the above-mentioned pull request (#1438), are:
`artifact.add_reference("wandb-artifact://<UPSTREAM_ARTIFACT_ID>/RESOURCE_PATH", "TARGET_PATH")`
and
`artifact.add_reference(other_artifact.get_path(<ASSET_NAME>), )`

Both are closer to what I have in mind while seeming still impractical for real deployment. Requiring proper artifact URIs for each individual image to be accessible at any point in a pipeline at which the user logs an image prediction attaches a massive burden to the designer. In ML/DL workflows we're often working with large batches of images stacked as raw numpy/tensor arrays with either no additional image info or simply a single integer label per-image. If we now need to keep track of wandb entity, project, artifact, and image URI locations for every one, the commitment necessary to really integrate wandb into these types of experiments may become higher than a programmer can justify.

Much of this is a recontextualized version of a comment I made on the wandb slack a week or two ago (~May 17th on channel #ask-for-feedback) which unfortunately did not land me any answers, reproduced below. 

> Another question regarding the new Dataset Visualization utility:
Topic: Regarding best practices to minimize redundant image data storage + only require upload of raw image once for many predictions
In the Image Classification with DSViz report (Really inspiring stuff btw), there's some lack of clarity around how exactly to organize image prediction logging while preventing the duplication of raw image data. Specifically, I think there would be immense value in getting some clarity targeted towards 2 distinct stages of the process, namely:
During an experiment, how to download a version-controlled image dataset from WandB -> mark as inputs for the run -> During prediction, the user must upload a wandb.Table with rows each containing a value for <???> in order to directly link the image used locally for prediction to an image stored on WandB's servers.
The tutorial has us upload both an id and the actual image during prediction, even though the documentation suggests only a shared id is required to benefit from the relational table comparison capabilities.
While allowing the user to upload the exact image used for prediction may help diagnose possible bugs introduced since downloading the images (e.g. the id may have become misaligned from its true image object), it also adds a significant increased demand in total data I/O over the network that requires the user to be more parsimonious with when to repeat predictions during training, or risk radically slowing down training.
From another perspective, this is a lost opportunity for maximally benefitting from the computational efficiency inherent to passing around image references (e.g. id) rather than the raw images themselves.
> 2. From the server perspective: How should we reference images during local artifact & table creation so that on the server, when inspecting our prediction tables, the image file referenced in each row correctly references an existing version-controlled image (possibly in another artifact), rather than creating a new reference to a new duplicate of the original image (e.g. the server containing 2 distinct files containing the image data for a single original sample, with each file possibly referenced by a different table).
Apologies for the wall of text. As I was typing this I realized it may be more suited as a GitHub issue or Pull request, so let me know if you think I should post it elsewhere. Regardless, I think clarity on this still somewhat abstracted aspect of the new data vis workflow would enable making it immensely more accessible. Thanks again!

**Comment by vanpelt** [comment]
Hey @JacobARose the best way to accomplish this is to use a [wandb.JoinedTable](https://docs.wandb.ai/guides/data-vis/tables#join-tables).  In this scenario your artifact structure would look like:

```
artifact 1  
     /media  
            /image1  
            /image2  
            ...
    source_table.json

artifact 2
    test_prediction_table.json
    test_prediction.joined-table.json
    /media
            /image1
            /image2
            ...
```

An example set of artifacts that leverages joined tables can be [found here](https://wandb.ai/stacey/cshanty/artifacts/analysis/synth_summary/0af204744b98ae0469cf/files).

Sometimes creating ID's for all of your examples and logging them to a table is non-trivial.  @tssweeney recently added the ability to join tables off of the digest of each blob in a media column.  He may be able to provide some more context / examples.

I think the key here is that you construct a `wandb.Table` in **artifact 1** so that you can always join those original rows to new predictions, let me know if I'm missing something here or you have additional questions.

**Comment by JacobARose** [comment]
@vanpelt Ahh thank you!! Enjoying the humor in the fact that, in the few hours since posting this issue, I had actually come across the webinar that you of all people presented some time last year that briefly described this exact strategy. I'm excited to try it out this week and will mention any other issues I run into on this page. I'm also eager to hear more detail about @tssweeney 's method of joining tables off the digest of each media blob if they get the chance.


Posting the link to the corresponding [blog post](https://wandb.ai/stacey/cshanty/reports/Visualize-Audio-Data-in-W-B--Vmlldzo1NDMxMDk) from that audio visualization wandb project you linked to. I think this + your webinar (Did you ever provide a link to the notebook from that webinar? I can post a link to the talk's info if necessary) should be sufficient guidance for now. After skimming back through some of the docs & examples, a couple preliminary questions/comments come to mind.

1. I must've missed the better JoinedTable example b/c that tutorial focuses exclusively on audio data which, despite being a musician, usually has very little relevance to my research in computer vision ^_^. I had previously gone through [this](https://colab.research.google.com/github/wandb/examples/blob/master/colabs/datasets-predictions/Image_Classification_with_DSViz.ipynb) excellent tutorial + project. My question:
    - Between steps 1 & 2, the author logs separate artifacts for (1) raw_data, then (2) split_data. While this is slightly different from the scenarios where we have a single referenced dataset on which we make+log multiple predictions, it also seems like a perfect candidate for deduplicating images in the same way. If users could create an arbitrary number of different data splits that only require the memory footprint of their corresponding json tables, that could be a massive benefit to basically every kind of DL/ML workflow that implements validation. In the past, I've often used git-version controlled csv tables for train/val/test or k-split validation that referenced the absolute paths of local image files, but that comes with the obvious flaw of providing opportunities for unobserved data changes, deletion or corruption, in addition to locking all deployment into a specific file system. Integrating the images directly into tables while deduplicating the raw image data itself would be quite an elegant alternative.
    - Since this tutorial doesn't make use of any deduplication benefits achieved from the use of wandb.JoinedTable, is there any plan for updating the tutorial, replacing it with a more up-to-date one on similar subject matter, or expanding the main documentation in this direction?


Thanks for any answers on these, so far this image predictions workflow seems to solve a bunch of frustrating engineering problems I've been forced to re-implement new solutions for far too many times during my PhD.

**Comment by tssweeney** [comment]
Hey @JacobARose: thanks for the engaging discussion. Perhaps I can share a bit more of how the underpinnings work which will shed light on some different options that can be applied.

1. Artifact Version Data Deduplication: If two versions of the same artifact have a file located in the same path, with the same sha256 digest, then only one copy is stored in the backend file store. This happens automatically on your behalf. So, let's say you have the example artifacts layout above, assuming that the images do not change in terms of their actual data, only one copy of each image is actually persisted. The main drawback here are that you still have to calculate the sha for potentially lots of images and this deduplication is only applied to versions across the same artifact.
2. Artifact References: If you choose to use `wandb.Image` (or other media types) to store your media assets, then we can intelligently track the source of such asset in the Python runtime object. This relates to number 4 in our writeup above (and 1.1 in your post above). So take the following example:
```
# Run 1:
run = wandb.init(...)
artifact = wandb.Artifact(..., ...)
artifact.add(wandb.Image(PATH_TO_LOCAL/NUMPY), "my_image")
run.log_artifact(artifact)

# Run 2
run = wandb.init(...)
art = run.use_artifact(...)
img = art.get("my_image") # this wandb.Image file knows where it came frome

art2 = wandb.Artifact(...)
art2.add(img, "my_image_2") # adding the runtime object fetched from artifact 1 into artifact 2
run.log_artifact(art2)
```
In this example, the data entry for "my_image_2" in the second artifact will show up as a blue link. It is a "Reference" to the entry in the first artifact. Now, as you pointed out, this is not practical for real production applications. So, we support this across table contents as well. (This is how the data split stuff works). So if instead, you are working with wandb.Tables, and those tables contain Images, then you can do something like:
```
art = run.use_artifact(...)
table = art.get("my_table_with_images") # assuming this has wandb.Images inside
art2 = wandb.Artifact(...)
table2 = wandb.Table(data=table.data[:len(table.data) // 2]) # notice I am just accessing part of the data. The internal wandb.Images know their source and therefore will build references like the example above automatically
art2.add(table2, "my_half_table")
run.log_artifact(art2)
```
In that examples, the new table's json structure would be persisted, but the actual image files would be references.
3. JoinedTables: JoinedTables allow you to create a composite table by joining two other tables on column. This is a good first step to building a normalized representation of your data. (Docs linked above)
4. Table Linking: Ok, so let's say you really want to build a normalized data structure. You can create foreign key references like in SQL. For example:
```
# Run 1: Create a dataset
run = wandb.init(...)
art = wandb.Artifact("dataset", "dataset")
tab = wandb.Table(...) # load this up with lots of great data
art.add(tab, "training_data")
run.log_artifact(art)

# Run 2: Create model predictions based on the data and create a linked table
run = wandb.init(...)
training = run.use_artifact("dataset:latest").get("training_data")
preds = wandb.Table(...) # load this up with predictions using "training" data
art = wandb.Artifact("prediction", "results")
# Assuming training has a column called `t_id` and `preds` has a column named `p_id` (just making them different for illustrative purposes)
preds.set_fk("p_id", training, "t_id") # makes and explicit link between the tables
art.add(preds, "pred_data")
run.log_artifact(art)
```
Notice in the example above, we do not re-log the training data/table/images at all - simply create a foreign key reference. In the UI, the tables will be automatically joined for your analysis. Note: this is more flexible than JoinedTable as this style of linking can be arbitrarily deep!

Ok, so that is everything to do with data deduplication. Hopefully this helps make a good decision for your use case - happy to chat some more.

As a bonus, let's take the above example. If you do not need full control over artifact naming, you can use an alternative notation to make the logging easier to write:

```
...
# Run 2: Create model predictions based on the data and create a linked table
run = wandb.init(...)
training = run.use_artifact("dataset:latest").get("training_data")
preds = wandb.Table(...) # load this up with predictions using "training" data
preds.set_fk("p_id", training, "t_id") # makes and explicit link between the tables
run.log({"my_predictions": preds})
```
Notice that there is no need to work with the lower level artifact objects in this case. You will still get an artifact in the artifact browser, however, containing your data.

Cheers, 
Tim
